containers/conformance.md

```diff
@@ -0,0 +1,220 @@
+      vendor="COMPANY NAME" \
+      version="VERSION NUMBER" \
+      release="RELEASE NUMBER" \
+      summary="APPLICATION SUMMARY" \
+      description="APPLICATION DESCRIPTION" \
+```
+
+### Tag
+
+Image tag can be a combination of these information in this format: `<version>-<linux-distro>[-sources][.<customize-build>]`.
+
+- **version**: it must follow [semantic versioning](https://semver.org/) or partial semantic versioning with major or major + minor. It may also be `latest` or `lts`. For example, `1`, `1.23`, `1.23.0`, `lts`, `latest`, etc.
+- **linux-distro**: for example, `ubi`, `ubuntu`, etc.
+- **customize-build**: string sanitized by converting non-letters and non-digits to dashes. For example, `pr-1234`, `users-john-fix123`, etc.
+- **Source Build**: must be string `-sources` appended to the end of tag.
+  * If this is a source build, the tag must contain full version number (major+minor+patch) information.
+  * Linux Distro information is recommended.
+  * Must NOT contain customize build information.
+  * For example: `1.23.0-ubi-sources`.
+
+For example, these are valid image tags:
+
+- latest
+- latest-ubuntu
+- latest-ubuntu-sources
+- latest-ubi
+- latest-ubi-sources
+- lts
+- lts-ubuntu
+- lts-ubi
+- 1
+- 1-ubuntu
+- 1-ubi
+- 1.23
+- 1.23-ubuntu
+- 1.23-ubi
+- 1.23.0
+- 1.23.0-ubuntu
+- 1.23.0-ubuntu-sources
+- 1.23.0-ubi
+- 1.23.0-ubi-sources
+- 1.23.0-alpine.pr-1234
+- 1.23.0-ubi.users-john-test1
+
+### Files and Directories
+
+This is required folder structure for all Zowe components:
+
+```
+/licenses
+/app
+  +- README.md
+```
+
+- `/licenses` folder holds all license related files. It should include at least the license information for current application. It's recommended to include license notice file of all pedigree dependencies. All licenses files must be in UTF-8 encoding.
+- `/app/README.md` provides information about the application for end-user.
+
+These file(s) and folder(s) are recommended:
+
+```
+/app
+  +- manifest.json or manifest.yaml
+  +- /bin/<lifecycle-scripts>
+  +- <other-application-files>
+```
+
+- `/app/manifest.(json|yaml)` is recommended for Zowe component. The format of the file is defined at [Zowe component manifest](https://docs.zowe.org/stable/extend/packaging-zos-extensions/#zowe-component-manifest). Component must use same manifest file when it's running on z/OS.
+- `/app/bin/<lifecycle-scripts>` should remain same as what it is running on z/OS. Component must use same lifecycle scripts when it's running on z/OS.
+
+### Environment Variable(s)
+
+These environment variable(s) must be set as a fixed value in the image:
+
+- `ZOWE_COMPONENT_ID`: this is the Zowe component ID for current image. For example: `ENV ZOWE_COMPONENT_ID=gateway`.
+
+### User `zowe`
+
+In the Dockerfile, a `zowe` user and group must be created. Example command `RUN groupadd -r zowe && useradd --no-log-init -r -g zowe zowe`.
+
+`USER zowe` must be specified before the first `CMD` or `ENTRYPOINT`.
```

Comment:
i had a problem with doing this in tech preview. volume mounts had certain user:group permissions that either required the person running the container to have matched the uid/gid up with what the container would have OR have the container initially start as root, reassign the container's zowe uid to be within (from container viewpoint) the gid seen in the volume, and THEN switch to user zowe to run zowe. therefore, i couldnt do `USER zowe` but did run zowe as user zowe, after initialization as root.

Reply:
This may cause trouble if we want to get certification. Will keep this in mind and try if I can solve the issue.

The issue could be caused by in tech preview, the workspace volume is touched off container. We should avoid that. All changes to persistent volume should come from container.

But we may need to align uid and gid for all images. Will add this part.

Reply:
unless there's some special utility i am not aware of, what i noticed is that if you migrate a z/os instance to linux (and i'm working on a utility to make that easy) or even if you just make a new directory, that directory is owned by someone other than the docker user.
lets say, user "rob" with id=100, in group "dev", gid=200
on the linux system, you see `rwxrwxr-x rob:dev instance`
in docker, you see `rwxrwxr-x 100:200 instance` but the zowe user is for example, id=1001 in gid=1002
so, the zowe user wouldnt be able to write to the instance, which can be a problem.

in that scenario, the only solution i found was to tell the user to manually `chmod` or `chown` to have the uid/gid match a known and hardcoded container uid/gid. I had picked 8544 to match the app-server port, and was going to tell users to make sure to do a `chown -R youruser:8544 instance` before running... it could work, but automatic config is better, which is what i ended up doing by
1) startup as root
2) read gid
3) add zowe to group in container (no effect on host system, except that now zowe can read the instance it otherwise couldnt)
4) run as zowe